topdown/glob.go

```diff
@@ -47,18 +48,27 @@ func builtinGlobMatch(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Ter
 	}
 	id := builder.String()
 
-	globCacheLock.Lock()
-	defer globCacheLock.Unlock()
-	p, ok := globCache[id]
-	if !ok {
-		var err error
-		if p, err = glob.Compile(string(pattern), delimiters...); err != nil {
-			return err
+	// NOTE(philipc): We use an anonymous function here to ensure that the
+	// deferred mutex unlock happens before we hit the call to iter.
+	globCompileAndMatch := func() (bool, error) {
```

Comment:
I'm not sure I see the need to make this anonymous, since the mutex is a package level variable anyways.

Reply:
Elaborating on that, we could go with

```go
func globCompile(id, pattern string, delimiters []string) (glob.Pattern, error) {
		globCacheLock.Lock()
		defer globCacheLock.Unlock()
		p, ok := globCache[id]
		if !ok {
			var err error
			if p, err = glob.Compile(string(pattern), delimiters...); err != nil {
				return nil, err
			}
			globCache[id] = p
		}
		return p, nil
}
```

Reply:
The intent there was to leave the existing logic alone around `defer`, and just give it a sensible function boundary so that it runs the deferred mutex unlock before `iter()` is called. There are other ways this could be done; I'll see if I can come up with something cleaner.

Reply:
I chose to pull out that logic into a basic helper function. The function boundary is very clear then.